lib/agent-redirect.js

```diff
@@ -0,0 +1,210 @@
+const { InvalidArgumentError } = require('./core/errors')
+const util = require('./core/util')
+const EventEmitter = require('events')
+const { getGlobalAgent } = require('./agent')
+const assert = require('assert')
+
+const kMaxRedirections = Symbol('maxRedirections')
+const kAgent = Symbol('agent')
+
+class RedirectHandler {
+  constructor ({ agent, opts, handler }) {
+    this.agent = agent
+    this.location = null
+    this.abort = null
+    this.opts = opts
+    this.handler = handler
+    this.destroyed = false
+  }
+
+  destroy (err) {
+    this.destroyed = true
+
+    if (this.abort) {
+      this.abort()
+    }
+
+    this.handler.onError(err)
+  }
+
+  get (origin) {
+    return this[kAgent].get(origin)
+  }
+
+  onConnect (abort) {
+    if (this.destroyed) {
+      abort()
+    } else {
+      this.abort = abort
+      this.handler.onConnect(abort)
+    }
+  }
+
+  onUpgrade (statusCode, headers, socket) {
+    this.handler.onUpgrade(statusCode, headers, socket)
+  }
+
+  onError (error) {
+    this.handler.onError(error)
+  }
+
+  onHeaders (statusCode, headers, resume) {
+    if ([300, 301, 302, 303, 307, 308].indexOf(statusCode) === -1) {
```

Comment:
In node core I've seen this implemented using a Set and a `has` look up. It might be trivially faster than an array scan; however, this list only has 6 items so its probably a negligible perf difference (wouldn't be surprised if V8 optimizes this already since the list is constant)